torch/_dynamo/guards.py

```diff
@@ -468,11 +469,54 @@ def TENSOR_MATCH(self, guard: Guard):
                 for term in terms:
                     real_value = self.get(tensor_name + "." + term)
                     code.append(f"{tensor_name}.{term} == {real_value}")
-                self._produce_guard_code(guard, code)
             else:
                 self.tensor_check_names.append(tensor_name)
                 self.tensor_check_examples.append(value)
 
+            # A frame is valid for reuse with dynamic dimensions if the new dynamic dimensions are a
+            # strict subset of the old.
+            #
+            # The logic here is as follows:
+            #
+            # Every mark_dynamic directive is a user-knows-best command, which can incur a raise at tracing
+            # time if we find guards that run counter to the user directive.
+            #
+            # If the frame is compiled with any marked dynamic indices, let's call that set of indices X.
+            # When we evaluated inputs against the guards, given the same tensor with potentially new dynamic indices, let's call that set Y.
+            # Y must be a strict subset of X.
+            #
+            # This is the case because any newly introduced mark_dynamic directives have a chance of
+            # incurring raises, failing compilation. Any existing mark_dynamic indices that we lost are safe to lose
+            # as all it means is that we have gotten rid of a user directive which could incur a raise at compile time.
+            # In the case of when there is no Y, that is, there are no dynamic indices marked at all, the frame is safe to reuse
+            # as an empty set is a safe degeneration - that is, a strictly static tensor is always valid for a frame compiled with that same
+            # tensor + more onerous user directives.
+            static, message = tensor_shape_should_be_static(
+                value, guard.source, is_tensor=True
+            )
+            if not static:
+                if hasattr(value, "_dynamo_dynamic_indices"):
+                    # Note - Arguably - if a user marked_dynamic on a dim, but we have *no other* guards on it,
+                    # protecting the directive is inconsequential - and therefore accumulating this guard is spurious.
+                    # TODO(voz): Consider checking for if there are any references to this tensor's dim in guards
+                    # and only guard here if there are.
```

Comment:
But there will always be a reference to it, because if it's a 0/1 dim we will have guarded on that, and otherwise we will have guarded on it `>= 2`

Reply:
I still think it is a useful comment to keep around for when we have dynamic=False

Reply:
But when dynamic=False every dim will be checked???

Reply:
not with shape guards? Just tensor guards, right? Perhaps the comment should be *no other shape guards*

Reply:
but there's no shape guards in dynamic=False at all...

Reply:
That is what I mean, nevermind, I will remove it